include/ensmallen_bits/nsga2/nsga2_impl.hpp

```diff
@@ -409,47 +448,60 @@ inline bool NSGA2::Dominates(
   for (size_t i = 0; i < n_objectives; i++)
   {
     // P is worse than Q for the i-th objective function.
-    if (calculatedObjectives[candidateP](i) > calculatedObjectives[candidateQ](i))
+    if (calculatedObjectives[candidateP](i) >
+        calculatedObjectives[candidateQ](i))
       allBetterOrEqual = false;
 
     // P is better than Q for the i-th objective function.
-    else if (calculatedObjectives[candidateP](i) < calculatedObjectives[candidateQ](i))
+    else if (calculatedObjectives[candidateP](i) <
+             calculatedObjectives[candidateQ](i))
       atleastOneBetter = true;
   }
 
   return allBetterOrEqual && atleastOneBetter;
 }
 
 //! Assign crowding distance to the population.
-template <typename MatType>
-inline void NSGA2::CrowdingDistanceAssignment(
+template <typename MatType, typename ColType, typename CubeType>
+template <typename InputMatType, typename ObjectiveMatType>
+void NSGA2Type<MatType, ColType, CubeType>::CrowdingDistanceAssignment(
     const std::vector<size_t>& front,
-    std::vector<arma::Col<typename MatType::elem_type>>& calculatedObjectives,
-    std::vector<typename MatType::elem_type>& crowdingDistance)
+    std::vector<ObjectiveMatType>& calculatedObjectives,
+    std::vector<typename InputMatType::elem_type>& crowdingDistance)
 {
   // Convenience typedefs.
-  typedef typename MatType::elem_type ElemType;
+  typedef typename InputMatType::elem_type ElemType;
+  typedef typename ForwardType<MatType>::uvec UVecType;
 
   size_t fSize = front.size();
   // Stores the sorted indices of the fronts.
-  arma::uvec sortedIdx  = arma::regspace<arma::uvec>(0, 1, fSize - 1);
+  UVecType sortedIdx = regspace<UVecType>(0, 1, fSize - 1);
 
   for (size_t m = 0; m < numObjectives; m++)
   {
     // Cache fValues of individuals for current objective.
-    arma::Col<ElemType> fValues(fSize);
-    std::transform(front.begin(), front.end(), fValues.begin(),
-      [&](const size_t& individual)
-        {
-          return calculatedObjectives[individual](m);
-        });
+    ColType fValues(fSize);
+    for (size_t i = 0; i < front.size(); ++i)
+    {
+      size_t individual = front[i];
+      fValues[i] = calculatedObjectives[individual](m);
+    }
 
     // Sort front indices by ascending fValues for current objective.
-    std::sort(sortedIdx.begin(), sortedIdx.end(),
-      [&](const size_t& frontIdxA, const size_t& frontIdxB)
+    for (size_t i = 0; i < sortedIdx.size(); ++i)
+    {
+      size_t minIdx = i;
+      for (size_t j = i + 1; j < sortedIdx.size(); ++j)
+      {
+        if (fValues(sortedIdx[j]) < fValues(sortedIdx[minIdx]))
         {
-          return (fValues(frontIdxA) < fValues(frontIdxB));
-        });
+          minIdx = j;
+        }
+      }
+      size_t temp = sortedIdx[i];
+      sortedIdx[i] = sortedIdx[minIdx];
+      sortedIdx[minIdx] = temp;
+    }
```

Comment:
This sorting (the entire block above) is done on the CPU with `std::sort`; but couldn't we use `sort_index()` here instead to achieve the same result?